### ISSUE TITLE
Parse error with variable named "pattern"

### ISSUE THREAD

**Issue by brandon-leapyear** [ndmitchell/hlint#607]
I can't find a minimal example (sorry!), but I have this code:
```haskell
foo = do
  ...
  matches <- reify name >>= \case
    TyConI (DataD _ _ _ _ cons _) -> return $ flip map cons $ \case
      NormalC con [] ->
        let pattern = litP . stringL . map toLower . nameBase $ con
        in match pattern (normalB [| pure $(conE con) |]) []
      con -> fail $ "Invalid constructor: " ++ show con
    info -> fail $ "Invalid data type: " ++ show info
  ...
```
which compiles fine, but hlint gives me
```
./aeson-schemas/src/Data/Aeson/Schema/TH/Enum.hs:49:23: Error: Parse error
Found:
          TyConI (DataD _ _ _ _ cons _) -> return $ flip map cons $ \case
            NormalC con [] ->
  >           let pattern = litP . stringL . map toLower . nameBase $ con
              in match pattern (normalB [| pure $(conE con) |]) []
            con -> fail $ "Invalid constructor: " ++ show con
  

1 hint
```

**Comment by brandon-leapyear** [comment]
Never mind, got it:
```haskell
main = let pattern = 1 in return ()
```
hlint doesn't like variables named `pattern`

**Comment by ndmitchell** [comment]
You can fix this by passing `-XNoPatternSymonyms`, which disables the extension that is problematic.

**Comment by brandon-leapyear** [comment]
I mean, as a workaround, I just renamed the variable. But I never had the extension turned on, and I don't think hlint should have it on by default; `pattern` is not an unreasonable name to give to a variable

**Comment by ndmitchell** [comment]
It's a question of do more programs turn on pattern synonyms and not write it in their extension, or does it break more programs to turn it on by default. I think for `pattern` its probably worth having it on by default. If you want nothing on by default pass `-XHaskell2010` to HLint and it won't do anything as standard.

**Comment by brandon-leapyear** [comment]
Hm I can understand that philosophy. I think I probably just ascribe to stylish-haskell's philosophy more, in the sense that they parse extensions from the file, and if you have extensions turned on in the cabal file, you need to explicitly tell stylish-haskell to turn on extensions; otherwise, no extensions are turned on.

This behavior makes sense within your philosophy of hlint, and I'm not trying to change that philosophy, so feel free to close this issue.

**Comment by ndmitchell** [comment]
Yep, different approaches. Partly because people commit to stylish haskell strongly, whereas people use HLint less seriously. Closing.
